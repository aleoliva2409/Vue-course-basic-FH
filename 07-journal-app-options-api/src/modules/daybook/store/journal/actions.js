import journalApi from '@/api/journalApi';
// export const actions = async ({ commit }) => {

// }

export const loadEntries = async ({ commit }) => {
  try {
    const { data } = await journalApi.get('/entries.json');

    if(!data) return commit('setEntries', [])

    const entries = [];
    for (let id of Object.keys(data)) {
      entries.push({
        id,
        ...data[id],
      });
    }

    commit('setEntries', entries);
  } catch (error) {
    console.log(error);
  }
};

export const updateEntry = async ({ commit }, entry) => {
  try {
    // const {id, text, date, picture} = entry
    // const { data } = await journalApi.put(`/entries/${id}`, {
    //   text,
    //   date,
    //   picture,
    // });
    const { id, ...rest } = entry;
    await journalApi.put(`/entries/${id}.json`, { ...rest });

    commit('updateEntry', { ...entry });
  } catch (error) {
    console.log(error);
  }
};

export const createEntry = async ({ commit }, entry) => {
  try {
    const { text, date, picture } = entry;

    const { data } = await journalApi.post('/entries.json', {
      text,
      date,
      picture,
    });

    commit('addEntry', { text, date, picture, id: data.name });

    return data.name; // ? id
  } catch (error) {
    console.log(error);
  }
};

export const deleteEntryById = async ({ commit }, id) => {
  try {
    await journalApi.delete(`/entries/${id}.json`)

    commit('removeEntry', id)

  } catch (error) {
    console.log(error);
  }
};
