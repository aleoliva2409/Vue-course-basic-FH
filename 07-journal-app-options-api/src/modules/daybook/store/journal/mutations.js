// export const mutations = ( state ) => {
// }

export const setEntries = (state, entries) => {
  state.entries = [...state.entries, ...entries];
  state.isLoading = false;
};

export const updateEntry = (state, entryUpdated) => {
  const indexEntry = state.entries
    .map((entry) => entry.id)
    .indexOf(entryUpdated.id);
  state.entries[indexEntry] = entryUpdated;
};

export const addEntry = (state, entry) => {
  state.entries = [entry, ...state.entries];
};

export const removeEntry = (state, id) => {
  state.entries = state.entries.filter(entry => entry.id !== id);
};
