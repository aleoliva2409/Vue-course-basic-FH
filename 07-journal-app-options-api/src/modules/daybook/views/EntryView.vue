<template v-if="entry">
  <div class="entry-title d-flex justify-content-between p-2">
    <div>
      <span class="text-success fs-3 fw-bold">{{ day }}</span>
      <span class="mx-1 fs-3">{{ month }}</span>
      <span class="mx-2 fs-4 fw-light">{{ year }}</span>
    </div>
    <div>
      <input
        type="file"
        @change="selectedImg"
        ref="imgSelector"
        v-show="false"
        accept="image/png, image/jpg, image/jpeg"
      />
      <button v-if="entry.id" @click="deleteEntry" class="btn btn-danger m-1">
        Borrar<i class="fa fa-trash-alt mx-1"></i>
      </button>
      <button class="btn btn-primary m-1" @click="selectImg">
        Subir foto<i class="fa fa-upload mx-2"></i>
      </button>
    </div>
  </div>
  <hr />
  <div class="d-flex flex-column px-3 h-75">
    <textarea v-model="entry.text" placeholder="Que sucedio hoy?"></textarea>
  </div>

  <Fab icon="fa-save" @on-click="saveEntry" />

  <img v-if="entry.picture && !localImg" :src="entry.picture" alt="lol" class="img-thumbnail" />

  <img v-if="localImg" :src="localImg" alt="lol" class="img-thumbnail" />
</template>

<script>
import { mapActions, mapGetters } from 'vuex';
import Fab from '../components/Fab.vue';
import { getDayMonthYear } from '../helpers/getDayMonthYear';
import { uploadImages } from '../helpers/uploadImages'

export default {
  components: {
    Fab,
  },
  props: {
    id: {
      type: String,
      required: true,
    },
  },
  computed: {
    ...mapGetters('journal', ['getEntryById']),
    // ...mapActions('journal', ['updateEntry']),
    day() {
      const { day } = getDayMonthYear(this.entry.date);
      return day;
    },
    month() {
      const { month } = getDayMonthYear(this.entry.date);
      return month;
    },
    year() {
      const { year } = getDayMonthYear(this.entry.date);
      return year;
    },
  },
  data() {
    return {
      entry: null,
      localImg: null,
      file: null,
    };
  },
  methods: {
    ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntryById']), //ahi lo ponen pero yo lo puse arriba
    loadEntry() {
      let entry;
      this.localImg = null;
      this.file = null;

      if (this.id === 'new') {
        entry = {
          text: '',
          date: new Date().getTime(),
        };
      } else {
        entry = this.getEntryById(this.id);
        if (!entry) return this.$router.push({ name: 'no-entry' });
      }

      this.entry = entry;
    },
    async saveEntry() {

      this.$swal({
        title: 'Espere por favor',
        allowOutsideClick: false
      })

      this.$swal.showLoading()

      const picture = await uploadImages(this.file)
      this.entry.picture = picture

      if (this.entry.id) {
        await this.updateEntry(this.entry);
      } else {
        const id = await this.createEntry(this.entry);
        this.$router.push({ name: 'entry', params: { id } });
      }

      this.file = null
      this.$swal('Guardado', 'Entrada registrada con exito', 'success');
    },

    async deleteEntry() {
      const { isConfirmed } = await this.$swal({
        title: 'Esta seguro?',
        text: 'Una vez borrado, no se podra recuperar',
        showDenyButton: true,
        confirmButtonText: 'Si, estoy seguro',
      });

      if (this.entry.id && isConfirmed) {
        this.$swal({
          title: 'Espere por favor',
          allowOutsideClick: false,
        });

        this.$swal.showLoading();
        await this.deleteEntryById(this.entry.id);
        this.$router.push({ name: 'no-entry' });

        this.$swal('Eliminado', '', 'success');
      }
    },
    selectedImg(e) {
      const imgFile = e.target.files[0];

      if (!imgFile) {
        this.file = null
        this.localImg = null
        return
      }

      this.file = imgFile;
      const fr = new FileReader();
      fr.onload = () => (this.localImg = fr.result);
      fr.readAsDataURL(imgFile);
    },
    selectImg() {
      this.$refs.imgSelector.click();
    },
  },
  watch: {
    id() {
      this.loadEntry();
    },
  },
  created() {
    this.loadEntry();
  },
};
</script>

<style lang="scss" scoped>
textarea {
  font-size: 20px;
  border: none;
  height: 100%;

  &:hover {
    outline: none;
  }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0 5px 10px rgba($color: #000000, $alpha: 0.2);
}
</style>
