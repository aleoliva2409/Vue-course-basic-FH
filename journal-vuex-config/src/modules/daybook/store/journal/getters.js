// export const getters = ( state ) => {
//   return state
// }


export const getEntries = ( state ) => {
  return state
}


export const getEntryById = ( state ) => {
  return state
}
