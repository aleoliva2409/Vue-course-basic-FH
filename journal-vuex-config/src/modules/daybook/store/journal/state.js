export default () => ({
  isLoading: true,
  entries: [
    {
      id: new Date().getTime(),
      date: new Date().toDateString(),
      text: 'Lorem, ipsum dolor sit amet consectetur adipisicing elit. Vero adipisci aliquam recusandae magnam excepturi deleniti quos architecto aliquid repellendus fugit!',
      picture: null
    },
    {
      id: new Date().getTime() + 1000,
      date: new Date().toDateString(),
      text: 'Lorem, ipsum dolor sit amet consectetur adipisicing elit. Vero adipisci aliquam recusandae magnam excepturi deleniti quos architecto aliquid repellendus fugit!',
      picture: null
    },
    {
      id: new Date().getTime() + 2000,
      date: new Date().toDateString(),
      text: 'Lorem, ipsum dolor sit amet consectetur adipisicing elit. Vero adipisci aliquam recusandae magnam excepturi deleniti quos architecto aliquid repellendus fugit!',
      picture: null
    },
    {
      id: new Date().getTime() + 3000,
      date: new Date().toDateString(),
      text: 'Lorem, ipsum dolor sit amet consectetur adipisicing elit. Vero adipisci aliquam recusandae magnam excepturi deleniti quos architecto aliquid repellendus fugit!',
      picture: null
    },
    {
      id: new Date().getTime() + 4000,
      date: new Date().toDateString(),
      text: 'Lorem, ipsum dolor sit amet consectetur adipisicing elit. Vero adipisci aliquam recusandae magnam excepturi deleniti quos architecto aliquid repellendus fugit!',
      picture: null
    }

  ]
})