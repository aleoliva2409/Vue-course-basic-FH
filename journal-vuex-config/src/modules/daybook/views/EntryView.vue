<template>
  <div class="entry-title d-flex justify-content-between p-2">
    <div>
      <span class="text-success fs-3 fw-bold">30</span>
      <span class="mx-1 fs-3">noviembre</span>
      <span class="mx-2 fs-4 fw-light">2021</span>
    </div>
    <div>
      <button class="btn btn-danger m-1">
        Borrar<i class="fa fa-trash-alt mx-1"></i>
      </button>
      <button class="btn btn-primary m-1">
        Subir foto<i class="fa fa-upload mx-2"></i>
      </button>
    </div>
  </div>
  <hr>
  <div class="d-flex flex-column px-3 h-75">
    <textarea placeholder="Que sucedio hoy?" ></textarea>
  </div>

  <Fab icon="fa-save" />

  <img src="https://i.blogs.es/bf5ad0/league-of-legends-destacada/1366_2000.jpeg" alt="lol" class="img-thumbnail">
</template>

<script>
import Fab from '../components/Fab.vue'

export default {
  components: {
    Fab
  }
};
</script>

<style lang="scss" scoped>
textarea {
  font-size: 20px;
  border: none;
  height: 100%;

  &:hover {
    outline: none;
  }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0 5px 10px rgba($color: #000000, $alpha: 0.2);
}

</style>
