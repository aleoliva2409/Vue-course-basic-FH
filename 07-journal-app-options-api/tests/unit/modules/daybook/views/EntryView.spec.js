import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';

// import Swal from 'sweetalert2';

import journal from '@/modules/daybook/store/journal';
import { journalState } from '../../../mocks/test-journal-state';

import EntryView from '@/modules/daybook/views/EntryView.vue';

// !! Problemas con sweetalert2, no pasan los tests

const createVuexStore = (initialState) =>
  createStore({
    modules: {
      journal: {
        ...journal,
        state: { ...initialState },
      },
    },
  });

// jest.mock('sweetalert2', () => ({
//   fire: jest.fn(),
//   showLoading: jest.fn(),
//   close: jest.fn(),
// }));

describe('Tests on EntryView', () => {
  const store = createVuexStore(journalState);
  store.dispatch = jest.fn();

  const mockRouter = {
    push: jest.fn(),
  };

  const mockSwal = {
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn(),
  };

  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallowMount(EntryView, {
      props: {
        id: 'ABC123',
      },
      global: {
        mocks: {
          $router: mockRouter,
          $swal: mockSwal,
        },
        plugins: [store],
      },
    });
  });

  xtest('debe de sacar al usuario porque el id no existe', () => {
    shallowMount(EntryView, {
      props: {
        id: 'Este ID no existe en el STORE',
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store],
      },
    });

    expect(mockRouter.push).toHaveBeenCalledWith({ name: 'no-entry' });
  });

  xtest('debe de mostrar la entrada correctamente', () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(mockRouter.push).not.toHaveBeenCalled();
  });

  xtest('debe de borrar la entrada y salir', (done) => {
    mockSwal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }));

    wrapper.find('.btn-danger').trigger('click');

    expect(mockSwal.fire).toHaveBeenCalledWith({
      title: 'Esta seguro?',
      text: 'Una vez borrado, no se podra recuperar',
      showDenyButton: true,
      confirmButtonText: 'Si, estoy seguro',
    });

    setTimeout(() => {
      expect(store.dispatch).toHaveBeenCalledWith(
        'journal/deleteEntry',
        'ABC123'
      );
      expect(mockRouter.push).toHaveBeenCalled();
      done();
    }, 1);
  });
});
