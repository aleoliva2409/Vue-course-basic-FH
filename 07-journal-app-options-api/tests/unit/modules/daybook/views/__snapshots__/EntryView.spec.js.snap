// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on EntryView debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">23</span><span class="mx-1 fs-3">Julio</span><span class="mx-2 fs-4 fw-light">2021</span></div>
  <div><input type="file" accept="image/png, image/jpg, image/jpeg" style="display: none;"><button class="btn btn-danger m-1"> Borrar<i class="fa fa-trash-alt mx-1"></i></button><button class="btn btn-primary m-1"> Subir foto<i class="fa fa-upload mx-2"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="Que sucedio hoy?"></textarea></div>
<fab-stub icon="fa-save"></fab-stub>
<img src="https://res.cloudinary.com/dx0pryfzn/image/upload/v1627077252/ip3vwn8sjvnncphqka4y.jpg" alt="lol" class="img-thumbnail">
<!--v-if-->
`;
